helix-core/src/shellwords.rs

```diff
@@ -1,198 +1,476 @@
+                        // e.g: :yank-join , "
+                        self.is_finished = true;
+                        return Some(&self.input[self.idx..self.bytes.len()]);
+                    } else if self.start < self.idx && !is_escaped(&self.bytes[..self.idx]) {
+                        // When part of the input end in a quote, `one two" three` this returns the `two` properly.
+                        let arg = Some(&self.input[self.start..self.idx]);
+                        self.start = self.idx;
+                        return arg;
+                    } else if self.idx == 0 || !is_escaped(&self.bytes[..self.idx]) {
+                        self.in_quotes = true;
+                        self.quote = self.bytes[self.idx];
+                        self.idx += 1;
+                        // Exclude quote from arg output.
+                        self.start = self.idx;
+
+                        // Check if quote is ever closed, and if not, then return the rest of the input as one arg.
+                        let mut found = false;
+                        for idx in self.start..self.input.len() {
+                            if self.bytes[idx] == self.quote && !is_escaped(&self.bytes[..self.idx])
+                            {
+                                found = true;
+                                break;
+                            }
+                        }
+
+                        if !found {
+                            self.is_finished = true;
+                            return Some(&self.input[self.idx..]);
+                        }
+                    } else {
+                        self.idx += 1;
+                    }
+                }
+                b' ' | b'\t' if !self.in_quotes && !self.in_list => {
+                    if self.idx + 1 == self.bytes.len() {
+                        self.is_finished = true;
+                        // Preserves whitespace if very last char of input.
+                        return Some(&self.input[self.start..=self.idx]);
+                    } else if self.start < self.idx {
+                        let arg = Some(&self.input[self.start..self.idx]);
+                        self.idx += 1;
+                        self.start = self.idx;
+                        return arg;
+                    }
+                    self.idx += 1;
+                    self.start = self.idx;
+                }
+                b'[' if !self.in_quotes => {
+                    self.in_list = true;
+                    // Includes the `[` bracket
+                    self.start = self.idx;
+                    self.idx += 1;
+                }
+                b']' if self.in_list => {
+                    self.in_list = false;
+                    // Includes the `]` bracket
+                    let arg = Some(&self.input[self.start..=self.idx]);
+                    self.idx += 1;
+                    self.start = self.idx;
+                    return arg;
+                }
```

Comment:
Indeed, this was a hold over from the previous functionality. And that sounds good to me. I can handle a "special case" equivalent to `raw` for `Args`, `rest() -> &str` , that would be `self.idx` aware, so that after a `next` to get the key, in this example, it can spit out the remaining in the same was as `raw`, so that there is no need for an implementer to worry about the text being preprocessed in same way that the interpreter couldn't handle(in this example).